File: cli/lib/gitignore.js
/**
 * Gitignore management — proper entry-by-entry handling.
 *
 * Fixes the bug where template literal \n gets written literally,
 * mashing all entries onto one line.
 */
import fs from 'fs';
import path from 'path';

const MARKER = '# mint local state';

/** Entries that mint needs gitignored in every project. */
export const MINT_IGNORE_ENTRIES = [
  '.mint/tasks/',
  '.mint/research/',
  '.mint/worktrees/',
  '.mint/plugins/',
  '.mint/sessions/',
  '.mint/ssh-cache.json',
  '.mint/.freeze-list.json',
  '.mint/.browser-sessions.json',
  '.mint/.gate-ledger.jsonl',
  '.mint/.dream-lock',
  '.mint/dream-report.md',
  '.mint/issues-archive.jsonl',
  '.mint/wins-archive.jsonl',
];

/**
 * Ensure a list of entries exist in .gitignore.
 * Adds missing entries one per line under the mint marker section.
 * Creates .gitignore if it doesn't exist.
 *
 * @param {string} projectRoot - project root directory
 * @param {string[]} [entries] - entries to ensure (defaults to MINT_IGNORE_ENTRIES)
 * @returns {{ added: string[], alreadyPresent: string[] }}
 */
export function ensureGitignore(projectRoot, entries = MINT_IGNORE_ENTRIES) {
  const gitignorePath = path.join(projectRoot, '.gitignore');

  let content = '';
  try { content = fs.readFileSync(gitignorePath, 'utf8'); } catch { /* no .gitignore yet */ }

  const added = [];
  const alreadyPresent = [];

  for (const entry of entries) {
    if (content.includes(entry)) {
      alreadyPresent.push(entry);
    } else {
      added.push(entry);
    }
  }

  if (added.length === 0) return { added, alreadyPresent };

  // Build the block to append
  const lines = [];

  // Add marker if not present
  if (!content.includes(MARKER)) {
    lines.push('', MARKER);
  }

  for (const entry of added) {
    lines.push(entry);
  }

  const block = lines.join('\n') + '\n';
  fs.appendFileSync(gitignorePath, block);

  return { added, alreadyPresent };
}

File: cli/internal/gitignore/gitignore.go
package gitignore

import (
	"os"
	"path/filepath"
	"strings"
)

const marker = "# mint local state"

var MintIgnoreEntries = []string{
	".mint/tasks/",
	".mint/worktrees/",
	".mint/sessions/",
	".mint/verdicts/",
	"scratchpad/",
}

type Result struct {
	Added          []string
	AlreadyPresent []string
}

func Ensure(projectRoot string, entries []string) (Result, error) {
	if entries == nil {
		entries = MintIgnoreEntries
	}

	gitignorePath := filepath.Join(projectRoot, ".gitignore")
	b, err := os.ReadFile(gitignorePath)
	if err != nil && !os.IsNotExist(err) {
		return Result{}, err
	}
	content := string(b)

	var result Result
	for _, entry := range entries {
		if strings.Contains(content, entry) {
			result.AlreadyPresent = append(result.AlreadyPresent, entry)
		} else {
			result.Added = append(result.Added, entry)
		}
	}

	if len(result.Added) == 0 {
		return result, nil
	}

	var lines []string
	if !strings.Contains(content, marker) {
		lines = append(lines, "", marker)
	}
	lines = append(lines, result.Added...)
	block := strings.Join(lines, "\n") + "\n"

	f, err := os.OpenFile(gitignorePath, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0o644)
	if err != nil {
		return Result{}, err
	}
	defer f.Close()
	if _, err := f.WriteString(block); err != nil {
		return Result{}, err
	}
	return result, nil
}
